apps/src/templates/progress/progressHelpers.js

```diff
@@ -151,87 +148,73 @@ export function progressForLesson(studentLevelProgress, levels) {
-    0;
+    aggregates.attempted + aggregates.imperfect + aggregates.completed > 0;
 
-  const getPercent = count => (100 * count) / statuses.length;
+  const getPercent = count => (100 * count) / filteredLevels.length;
   return {
     isStarted: isLessonStarted,
-    imperfectPercent: getPercent(statusCounts.imperfect),
-    completedPercent: getPercent(statusCounts.completed),
-    incompletePercent: getPercent(incomplete)
+    incompletePercent: getPercent(incomplete),
+    imperfectPercent: getPercent(aggregates.imperfect),
+    completedPercent: getPercent(aggregates.completed),
+    timeSpent: aggregates.timeSpent,
+    lastTimestamp: aggregates.lastTimestamp
   };
 }
 
 /**
- * Summarizes stage progress data.
- * @param {{id:studentLevelProgressType}} studentProgress An object keyed by
- * level id containing objects representing the student's progress in that level
- * @param {levelType[]} levels An array of the levels in a stage
- * @returns {object} An object with a total count of levels in each of the
- * following buckets: total, completed, imperfect, incomplete, attempted.
+ * Computes studentLessonProgressType objects for each lesson from the provided
+ * level progress data for each student.
+ * @param {studentId: {levelId: studentLevelProgressType}} sectionLevelProgress
+ * An object keyed by student id all the student's level progress data
+ * @param {lessonType[]} lessons An array of lessons
+ * @returns {studentId: {lessonId: studentLessonProgressType}}
+ * An object containing lesson progress data for each student in a section
  */
-export function summarizeProgressInStage(studentProgress, levels) {
-  // Filter any bonus levels as they do not count toward progress.
-  const filteredLevels = levels.filter(level => !level.bonus);
-
-  // Get counts of statuses
-  let statusCounts = {
-    total: 0,
-    completed: 0,
-    imperfect: 0,
-    incomplete: 0,
-    attempted: 0
-  };
-
-  filteredLevels.forEach(level => {
-    const levelProgress = studentProgress[level.id];
-    statusCounts.total++;
-    if (!levelProgress) {
-      statusCounts.incomplete++;
-      return;
-    }
-    switch (levelProgress.status) {
-      case LevelStatus.perfect:
-      case LevelStatus.submitted:
-      case LevelStatus.free_play_complete:
-      case LevelStatus.completed_assessment:
-      case LevelStatus.readonly:
-        statusCounts.completed++;
-        break;
-      case LevelStatus.not_tried:
-        statusCounts.incomplete++;
-        break;
-      case LevelStatus.attempted:
-        statusCounts.incomplete++;
-        statusCounts.attempted++;
-        break;
-      case LevelStatus.passed:
-        statusCounts.imperfect++;
-        break;
-      // All others are assumed to be not tried
-      default:
-        statusCounts.incomplete++;
+export function lessonProgressForSection(sectionLevelProgress, lessons) {
+  const studentLessonProgress = {};
+  Object.entries(sectionLevelProgress).forEach(
```

Comment:
I wish there were a clearer way to express a map() over an object/hash in JS.  The operation here is conceptually straightforward but difficult (for me) to parse from this code.

Reply:
I also spent like 15 minutes trying to come up with a better way to express this, because I also found it confusing. But after 15 minutes determined that I couldn't think of a clearer way

Reply:
i'll add comments!